### DISCUSSION TITLE
How does Next.js compute the built-in geo localization data

### DISCUSSION THREAD

**Discussion by ovidiu-travelperk** [vercel/next.js#33445] [Help] [answered]
hello,

I'm trying to figure out how nextjs computes the geo info from `NextRequest.geo`, but had no success so far by just inspecting the source code.

is it based on data bundled with nextjs? is it calling an external geo IP service? does it have a quota?

trying to determine how reliable it is to use the provided geo data in a production environment

thank you

**Comment by joggienl** [comment, marked as answer]
My bett would be that it is being provisioned by the edge servers as a header once you an url.

**Reply by joggienl** [reply]
If you are using Vercel this might give you more information:

- https://vercel.com/changelog/ip-geolocation-for-serverless-functions
- https://vercel.com/changelog/ip-geolocation-now-available-for-all-plans
- https://vercel.com/docs/concepts/edge-network/headers#x-vercel-ip-country

So the serverless runtime on the edge is handling this, that is not part of nextjs itself but part of the vercel platform. I doubt that if you host a nextjs project elsewhere it would have those available.

I can’t answer the liability question though.

**Reply by leerob** [reply]
Made this more clear in the documentation, thank you https://github.com/vercel/next.js/pull/33609.
